### PR TITLE
Building a Multiclass image classification model using MobileNet-V2 and TensorFlow

### LINKED ISSUE #6758: [Machine Learning] Building a Multiclass image classification model using MobileNet-V2 and TensorFlow
#### Proposed title of article
Building a Multiclass image classification model with MobileNetV2 and TensorFlow

#### Proposed article introduction
Image classification categorizes input images into pre-defined labels or categories. Most image classifications usually have only two classes. In a multi-class, we have three or more classes such as `orange`, `mango`, `apple`, and `pineapple`. Multi-class image classification categorizes an input image into one of the three or more classes. A model is trained on a set of image datasets so that it can be able to classify images into pre-defined labels. Multiclass image classification is a common task in computer vision.

MobileNetV2 is a type of pre-trained convolutional neural network based on a streamlined architecture that uses convolutions to build light weight deep neural networks that can have low latency for mobile and embedded devices. MobileNets support any input size greater than 32 x 32, with larger image sizes offering better performance.

Pre-trained models are neural networks which are trained on very large datasets, developed, and are used to solve similar types of problems. We will use the pre-trained Mobile Net model to build a model that classifies various leaf images.

#### Key takeaways 
1. Application of transfer learning.
2. Downloading and customize the pre-trained MobileNet-v2.
3. Images standardization and scaling.
4. Fine-tuning the model.
5. Classification of various leaf images.
#### Article quality
This article is unique because it will explain how to download the pre-trained MobileNet-v2 from TensorFlow hub and customizing it to perform multi-class image classification. We also discuss how to fine-tune the model by adding custom Keras layer to improve the models performance. We will also discuss the important aspect of image preprocessing so that we provide the required image format for model training.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by kelivinnganga**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #6758

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+tfds.show_examples(info_train,train)
+```
+The code will display some of the train images as shown below:
+
+![Train images](/engineering-education/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/train-images.png)
+
+From this output, `Angular Leaf Spot` leaves are labeled 0, `Bean Rust` 1, and `healthy` 2. 
+
+### Image preprocessing
+Image preprocessing will convert the beans image dataset into a format the neural network can use. It involves various stages. Lets create a function that will perform [image scaling](https://en.wikipedia.org/wiki/Image_scaling), [normalization](https://en.wikipedia.org/wiki/Normalization_(image_processing)) and [one hot encoding](https://en.wikipedia.org/wiki/One-hot).
```
[the remark this answers is not in the document]
Comment (srishilesh) [reply content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:132]: `It involves various stages`

Explain a little about it. And, add any reference links

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+#### Function for image preprocessing
+We create this function using this code:
+
+```python
+def preprocessing(image, label):
+  image /= 255.0
+
+  return tf.image.resize(image,[224,224]), tf.one_hot(label, 3)
+```
+Image normalization converts the output pixel value between 0 by 1. The `preprocessing` function performs `normalization` by dividing the `image` by 255. The function uses ` tf.image.resize` to resize the image to `224,224`. It is the image size that MobileNetV2 expects.
```
Comment (srishilesh) [inline content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:143]: Use ` `` ` for keywords, variables, functions or anything that you want to be highlighted.

And, remove any additional spaces

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+The function also encodes the classes using `tf.one_hot(label, 3)`. It converts the classes (`Angular Leaf Spot`, `Bean Rust`, `healthy`), into numerical values (0, 1, 2)
```
Comment (srishilesh) [inline content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:145]: Why are we encoding the values? State the reason.

Add reference links at Section.

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+To use the pre-trained MobilenetV2 layers, use this code:
+
+```python
+mobile_net_layers = hub.KerasLayer(mobilenet_v2, input_shape=(224,224,3))
+```
+The code above will extract the pre-trained MobilenetV2 layers. It also specifies the size of the beans image we will feed the pre-trained MobilenetV2.
+
+`mobile_net_layers` will extract the unique characteristics and features from the beans images. We will then add our custom layers to the pre-trained MobilenetV2 using TensorFlow. It will produce a fine-tuned model that will classify the images into three classes.
+
+These layers are already trained. To ensure that TensorFlow does not retrain them, use this code:
```
Comment (srishilesh) [inline content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:191]: So, we are just using the pre-trained models to make predictions? If so, state that here. It's so not implicitly understandable.

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+From the code above, we have added the following layers to the deep neural network.
+
+**mobile_net_layers**
+It contains the pre-trained MobilenetV2 layers.
+
+**Dropout**
+This layer will handle [model overfitting](/engineering-education/dropout-regularization-to-handle-overfitting-in-deep-learning-models/). It ensures the model performs well using both the train and test images. 0.3% of the neurons will be dropped randomly in training to handle overfitting.
+
+**Dense**
+It is the output layer. The `Dense` layer has 3 neurons because the dataset has three predefined classes. We have also set `softmax` as the activation function because the dataset has three pre-defined classes.
```
Comment (srishilesh) [inline content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:220]: Make these bulleted points

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+It is an open-source library for developing machine learning and deep learning models. It trains deep neural networks with input, hidden, and output layers. We will use TensorFlow to add custom layers to the pre-trained MobilenetV2. This will fine-tune the plant disease classification model and improve its performance.
+
+**tensorflow_hub**
+It is an open-source repository that contains pre-trained models for natural language processing tasks and image classification. We will download the pre-trained MobilenetV2 model from here.
+
+**matplotlib.pylab**
+We will the library to plot diagrams and visualization of our image dataset. It will show the model prediction results.
+
+**numpy**
+It will convert the image dataset into an array.
```
Comment (srishilesh) [inline content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:169]: Make them as bulleted points.

File: content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+We compile the model using the `compile` function. It has the following parameters:
+
+**model_optimizer**
+It ensures the model performs well and reduces errors that `model_loss` generates in training. We set the `model_optimizer` to `Adam`.
+
+**model_loss**
+It gets the errors in the model in training. We set the `model_loss` to `CategoricalCrossentropy` because the beans dataset has three pre-defined classes.
+
+**model_metrics**
+It checks the deep neural network's performance. It also calculates the accuracy score. `acc` will get the accuracy score using the train and the test set.
```
Comment (srishilesh) [inline content/articles/building-a-multiclass-image-classification-model-using-mobilenet-v2-and-tensorflow/index.md:265]: Use bulleted points.

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @kelivinnganga ,

The tutorial looks great. It's easy to follow.

I've requested for some minor changes. Please let me know when you're done, we can then move to my final review.

Thanks.

**Comment by kelivinnganga** [comment]
Hello @srishilesh, I hade gone through all the suggestions and made the changes.

**Comment by srishilesh** [comment]
Let me check and get back to you by tomorrow.

**Comment by srishilesh** [comment]
I'll complete my final review by tomorrow.

**Review by srishilesh — APPROVED** [review]
Hey @kelivinnganga ,

Great article. The flow was good. You have covered all the topics.

Have a look at the changes I've made.

Moving for the final review.

Thanks.
